README.md
[![Circle CI](https://circleci.com/gh/valhalla/packaging.svg?style=svg)](https://circleci.com/gh/valhalla/packaging)

What is this?
-------------

Packaging for debian (ubuntu really). Check out `build.sh` to see how we are now packing valhalla into its library, devel and bin packages. It also has links and notes about how to get those packages onto your PPA. Enjoy and thanks!

What should I do?
-----------------

This was based off of the `prime_server` [packaging work](https://github.com/kevinkreiser/ppa-libprime-server) so your best bet is to head over there to see if that has relevant examples to get you up and running.


releases
trusty
xenial


build.sh
#!/bin/bash
set -e

#tell bzr who we are
DEBFULLNAME="valhalla"
DEBEMAIL="valhalla@mapzen.com"
bzr whoami "${DEBFULLNAME} <${DEBEMAIL}>"
source /etc/lsb-release

VERSION=$(cat version)
RELEASES=( $(cat releases) )
PACKAGE="$(if [[ "${1}" == "--versioned-name" ]]; then echo libvalhalla${VERSION}; else echo libvalhalla; fi)"

#boost!!!!!!
declare -A boost=( ["trusty"]="1.54" ["vivid"]="1.55" ["wily"]="1.58" ["xenial"]="1.58" )

#get a bunch of stuff we'll need to  make the packages
sudo apt-get install -y git dh-make dh-autoreconf bzr-builddeb pbuilder debootstrap devscripts distro-info

#get valhalla code into the form bzr likes
./prepare.sh ${VERSION} ${PACKAGE}
tar pczf ${PACKAGE}_${VERSION}.orig.tar.gz ${PACKAGE}
rm -rf ${PACKAGE}

######################################################
#LETS BUILD THE PACKAGE FOR SEVERAL RELEASES
for release in ${RELEASES[@]}; do
	rm -rf ${release}_build
	mkdir ${release}_build
	pushd ${release}_build

	#copy source targz
	cp -rp ../${PACKAGE}_${VERSION}.orig.tar.gz .
	tar pxf ${PACKAGE}_${VERSION}.orig.tar.gz

	#build the dsc and source.change files
	pushd ${PACKAGE}
	cp -rp ../../debian .
	if [[ "${1}" == "--versioned-name" ]]; then
		echo -e "libvalhalla${VERSION} (${VERSION}-0ubuntu1~${release}1) ${release}; urgency=medium\n" > debian/changelog
		for p in $(grep -F Package debian/control | sed -e "s/.*: //g"); do
			for ext in .dirs .install; do
				mv debian/${p}${ext} debian/$(echo ${p} | sed -e "s/valhalla/valhalla${VERSION}/g" -e "s/valhalla${VERSION}\([0-9]\+\)/valhalla${VERSION}-\1/g")${ext}
			done
		done
		sed -i -e "s/\([b| ]\)valhalla/\1valhalla${VERSION}/g" -e "s/valhalla${VERSION}\([0-9]\+\)/valhalla${VERSION}-\1/g" debian/control
	else
		echo -e "libvalhalla (${VERSION}-0ubuntu1~${release}1) ${release}; urgency=medium\n" > debian/changelog
	fi
	sed -i -e "s/BOOST_VERSION/${boost[${release}]}/g" debian/control
	curl https://raw.githubusercontent.com/valhalla/valhalla-docs/master/release-notes.md 2>/dev/null | sed -e "s/^##/*/g" -e "s/^\(.\)/  \1/g" >> debian/changelog
	echo -e "\n -- ${DEBFULLNAME} <${DEBEMAIL}>  $(date -u +"%a, %d %b %Y %T %z")" >> debian/changelog
	debuild -S -uc -sa
	popd

	#make sure we support this release
	if [ ! -e ~/pbuilder/${release}-base.tgz ]; then
		pbuilder-dist ${release} create	
	fi

	#try to build a package for it
	DEB_BUILD_OPTIONS="parallel=$(nproc)" pbuilder-dist ${release} build ${PACKAGE}_${VERSION}-0ubuntu1~${release}1.dsc --hookdir=../hooks
	popd
done
######################################################


prepare.sh
#!/bin/bash
set -e

VERSION=$(echo ${1} | sed -ne "/^[0-9]\+\.[0-9]\+\.[0-9]\+$/p")
DEST_DIR=${2:-libvalhalla}
REPOS="midgard baldr sif meili skadi mjolnir loki odin thor tyr tools"
DIRS="valhalla src test proto scripts m4 date_time lua locales"
source ./target
if [ -z ${VERSION} ]; then
	echo "$0 major.minor.patch"
	exit 1
fi

rm -rf ${DEST_DIR}
mkdir ${DEST_DIR}
pushd ${DEST_DIR}

#the version of the package
mkdir valhalla
cat > valhalla/valhalla.h << EOF
#ifndef __VALHALLA_VERSION_H__
#define __VALHALLA_VERSION_H__

#define VALHALLA_VERSION_MAJOR $(echo ${VERSION} | awk -F. '{print $1}')
#define VALHALLA_VERSION_MINOR $(echo ${VERSION} | awk -F. '{print $2}')
#define VALHALLA_VERSION_PATCH $(echo ${VERSION} | awk -F. '{print $3}')

#endif
EOF

#for pkg-config
cat > libvalhalla.pc.in << "EOF"
prefix=@prefix@
exec_prefix=@exec_prefix@
libdir=@libdir@
includedir=@includedir@

Name: libvalhalla
Description: valhalla c++ library
Version: @VERSION@
Libs: -L${libdir} -lvalhalla
Cflags: -I${includedir}
EOF

#generate configure script with this script
cat > autogen.sh << EOF
#!/bin/bash
aclocal -I m4
autoreconf -fi --warning=no-portability
EOF
chmod 755 autogen.sh

#copy in the good stuff
for r in ${REPOS}; do
	git clone --quiet --branch ${VERSION} --depth 1 --recursive  https://github.com/valhalla/${r}.git &
done
wait
for r in ${REPOS}; do
	for d in ${DIRS}; do
		if [ -e ${r}/${d} ]; then
			mkdir -p ${d}
			cp -rpL ${r}/${d}/* ${d}
		fi
	done
done

#some other autotools junk
cp -rp $(echo ${REPOS} | awk '{print $1}')/NEWS .
cp -rp $(echo ${REPOS} | awk '{print $1}')/AUTHORS .
cp -rp $(echo ${REPOS} | awk '{print $1}')/COPYING .
cat > README << "EOF"
See https://github.com/valhalla for more info about the various parts of the system.
EOF
curl 'https://raw.githubusercontent.com/valhalla/valhalla-docs/master/release-notes.md' 2>/dev/null > ChangeLog

cat > Makefile.am << EOF
ACLOCAL_AMFLAGS = -Im4
AM_LDFLAGS = $(grep -rE "^AM_LDFLAGS" */Makefile.am | sed -e "s/.*=\s*//g" -e "s/ /\n/g" | sort | uniq | tr '\n' ' ')
AM_CPPFLAGS = $(grep -rE "^AM_CPPFLAGS" */Makefile.am | sed -e "s/.*=\s*//g" -e "s/ /\n/g" | sort | uniq | tr '\n' ' ')
AM_CXXFLAGS = $(grep -rE "^AM_CXXFLAGS" */Makefile.am | sed -e "s/.*=\s*//g" -e "s/ /\n/g" | sort | uniq | tr '\n' ' ')
LIBTOOL_DEPS = @LIBTOOL_DEPS@
libtool: \$(LIBTOOL_DEPS)
	\$(SHELL) ./config.status libtool

# things for versioning
pkgconfig_DATA = libvalhalla.pc

# conditional test coverage
if ENABLE_COVERAGE
.PHONY: clean-coverage
clean-coverage:
	-find -name '*.gcda' -exec rm -rf {} \\;
	-\$(LCOV) --directory \$(top_builddir) -z
	-rm -rf coverage.info coverage/

.PHONY: coverage-report
coverage-report: clean-coverage
	-\$(MAKE) \$(AM_MAKEFLAGS) -k check
	\$(MAKE) \$(AM_MAKEFLAGS) coverage/index.html

coverage.info:
	\$(LCOV) --directory \$(top_builddir) --base-directory \$(top_builddir) --no-external --capture --output-file \$@ --no-checksum --compat-libtool

coverage/index.html: coverage.info
	\$(GENHTML) --prefix \$(top_builddir) --output-directory \$(@D) --title "Test Coverage" --legend --show-details $<


.PHONY: clean-gcno
clean-gcno:
	-find -name '*.gcno' -exec rm -rf {} \\;

clean-local: clean-coverage clean-gcno
endif

EOF


#extra targets date_time, lua, locales
extras=
for r in ${REPOS}; do
	for l in $(grep -nE "^src.*:" ${r}/Makefile.am | grep -vF proto | sed -e "s/:.*//g"); do
		target ${l} ${r}/Makefile.am >> Makefile.am
		extras="${extras} $(target ${l} ${r}/Makefile.am | head -n 1 | sed -e "s/:.*//g")"
	done
done

#proto targets
cat >> Makefile.am << EOF

PROTO_FILES = $(find proto/ -type f | tr '\n' ' ')
src/proto/%.pb.cc: proto/%.proto
	@echo " PROTOC $<"; mkdir -p src/proto valhalla/proto; @PROTOC_BIN@ -Iproto --cpp_out=valhalla/proto $< && mv valhalla/proto/\$(@F) src/proto

EOF

#add built stuff to sources, no dist and clean
cat >> Makefile.am << EOF
BUILT_SOURCES = \$(patsubst %.proto,src/%.pb.cc,\$(PROTO_FILES)) ${extras}
nodist_libvalhalla_la_SOURCES = \$(patsubst %.proto,src/%.pb.cc,\$(PROTO_FILES)) ${extras}
CLEANFILES = \$(patsubst %.proto,valhalla/%.pb.h,\$(PROTO_FILES)) \$(patsubst %.proto,src/%.pb.cc,\$(PROTO_FILES)) ${extras}

EOF

#lib and headers
#TODO: unclobber generated valhalla.h with midgards
echo "lib_LTLIBRARIES = libvalhalla.la
nobase_include_HEADERS = \\" >> Makefile.am
for r in ${REPOS}; do
	for l in $(grep -nE "^nobase_include_HEADERS" ${r}/Makefile.am | sed -e "s/:.*//g"); do
		target ${l} ${r}/Makefile.am | sed -e "s/[[:space:]\\\\]*$//g" -e "s/^\s*//g" -e "s/^nobase_include_HEADERS[ =]*//g" -e '/^\s*$/d'
	done
done | tr '\n' ' ' | sed -e "s/\s*$//g" -e "s/\s\+/ \\\\\n/g" | sed -e "s/^/\t/g" >> Makefile.am

#sources
echo "
libvalhalla_la_SOURCES = \\" >> Makefile.am
for r in ${REPOS}; do
	for l in $(grep -nE "^libvalhalla_${r}_la_SOURCES" ${r}/Makefile.am | sed -e "s/:.*//g"); do
		target ${l} ${r}/Makefile.am | sed -e "s/[[:space:]\\\\]*$//g" -e "s/^\s*//g" -e "s/^libvalhalla_${r}_la_SOURCES[ =]*//g" -e '/^\s*$/d'
	done
done | tr '\n' ' ' | sed -e "s/\s*$//g" -e "s/\s\+/ \\\\\n/g" | sed -e "s/^/\t/g" >> Makefile.am
echo >> Makefile.am

#flags
sed -e "s/^libvalhalla_.*_la_CPPFLAGS\s*=\s*//;tx;d;:x" */Makefile.am | tr ' ' '\n' | grep -vF VALHALLA | sort | uniq | tr '\n' ' ' | sed -e "s/^/libvalhalla_la_CPPFLAGS = /g" >> Makefile.am
echo >> Makefile.am

#libs to link
sed -e "s/^libvalhalla_.*_la_LIBADD\s*=\s*//;tx;d;:x" */Makefile.am | tr ' ' '\n' | grep -vF VALHALLA | sort | uniq | tr '\n' ' ' | sed -e "s/^/libvalhalla_la_LIBADD = /g" >> Makefile.am
echo >> Makefile.am
echo >> Makefile.am

#executable targets
echo "bin_PROGRAMS = \\" >> Makefile.am
bins=$(wc -l < Makefile.am)
for r in ${REPOS}; do
	for l in $(grep -nE "^bin_PROGRAMS" ${r}/Makefile.am | sed -e "s/:.*//g"); do
        	target ${l} ${r}/Makefile.am | sed -e "s/[[:space:]\\\\]*$//g" -e "s/^\s*//g" -e "s/^bin_PROGRAMS[ =]*//g" -e '/^\s*$/d'
        done
done | tr '\n' ' ' | sed -e "s/\s*$//g" -e "s/\s\+/ \\\\\n/g" | sed -e "s/^/\t/g" >> Makefile.am
echo >> Makefile.am
for e in $(target ${bins} Makefile.am | grep -vF bin_PROGRAMS | sed -e "s/[[:space:]\\]*//g"); do
	grep -hE "^${e}_[A-Z]" */Makefile.am | sed -e "s/libvalhalla_[a-z]\+\.la//g" -e "s/ [^[:space:]]\+VALHALLA_DEPS[^[:space:]]\+//g" -e "s/\(.*LDADD.*\)/\1 libvalhalla.la/g" >> Makefile.am
done
echo >> Makefile.am

#script targets
echo "bin_SCRIPTS = \\" >> Makefile.am
scripts=$(wc -l < Makefile.am)
for r in ${REPOS}; do
        for l in $(grep -nE "^bin_SCRIPTS" ${r}/Makefile.am | sed -e "s/:.*//g"); do
                target ${l} ${r}/Makefile.am | sed -e "s/[[:space:]\\\\]*$//g" -e "s/^\s*//g" -e "s/^bin_SCRIPTS[ =]*//g" -e '/^\s*$/d'
        done
done | tr '\n' ' ' | sed -e "s/\s*$//g" -e "s/\s\+/ \\\\\n/g" | sed -e "s/^/\t/g" >> Makefile.am
echo >> Makefile.am

#test targets
echo "TESTS_ENVIRONMENT = LOCPATH=locales" >> Makefile.am
echo "check_PROGRAMS = \\" >> Makefile.am
checks=$(wc -l < Makefile.am)
for r in ${REPOS}; do
	for l in $(grep -nE "^check_PROGRAMS" ${r}/Makefile.am | sed -e "s/:.*//g"); do
		target ${l} ${r}/Makefile.am | sed -e "s/[[:space:]\\\\]*$//g" -e "s/^\s*//g" -e "s/^check_PROGRAMS[ =]*//g" -e '/^\s*$/d'
	done
done | tr '\n' ' ' | sed -e "s/\s*$//g" -e "s/\s\+/ \\\\\n/g" | sed -e "s/^/\t/g" >> Makefile.am
echo >> Makefile.am
for t in $(target ${checks} Makefile.am | grep -vF check_PROGRAMS | sed -e "s/[[:space:]\\]*//g" -e "s@/@_@g"); do
	grep -hE "^${t}_[A-Z]" */Makefile.am | sed -e "s/libvalhalla_[a-z]\+\.la//g" -e "s/ [^[:space:]]\+VALHALLA_DEPS[^[:space:]]\+//g" -e "s/\(.*LDADD.*\)/\1 libvalhalla.la/g" >> Makefile.am
done
echo >> Makefile.am

cat >> Makefile.am << "EOF"
TESTS = $(check_PROGRAMS)
TEST_EXTENSIONS = .sh
SH_LOG_COMPILER = sh

test: check
EOF

#cleanup
rm -rf ${REPOS}

#TODO: create this from the subprojects' configure.ac's
cat > configure.ac << EOF
AC_INIT([valhalla],
	[${VERSION}],
	[https://github.com/valhalla/packaging/issues],
	[valhalla-${VERSION}],
	[https://github.com/valhalla/packaging])
AC_CONFIG_AUX_DIR([.])
AM_INIT_AUTOMAKE([subdir-objects parallel-tests])
LT_INIT
AC_SUBST([LIBTOOL_DEPS])

AM_SILENT_RULES([yes])
AC_CONFIG_HEADERS([valhalla/config.h])
AC_CONFIG_MACRO_DIR([m4])

# set pkgconfigdir, allow override
AC_ARG_WITH([pkgconfigdir],
            AS_HELP_STRING([--with-pkgconfigdir=PATH], [Path to the pkgconfig directory [[LIBDIR/pkgconfig]]]),
            [pkgconfigdir="\$withval"],
            [pkgconfigdir='\${libdir}/pkgconfig'])
AC_SUBST([pkgconfigdir])

AC_PROG_CXX
AC_PROG_INSTALL
AC_PROG_MAKE_SET

AC_HEADER_STDC
AC_LANG_CPLUSPLUS

# require c++11
AX_CXX_COMPILE_STDCXX_11([noext],[mandatory])

# check for protocol buffers compiler and libraries
REQUIRE_PROTOC

# check for boost and make sure we have the program options library
AX_BOOST_BASE([1.54], , [AC_MSG_ERROR([cannot find Boost libraries, which are are required for building valhalla. Please install libboost-all-dev.])])
AX_BOOST_PROGRAM_OPTIONS
AX_BOOST_SYSTEM
AX_BOOST_THREAD
AX_BOOST_FILESYSTEM
AX_BOOST_REGEX
AX_BOOST_DATE_TIME

# check for Lua libraries and headers
AX_PROG_LUA([5.2],[],[
    AX_LUA_HEADERS([
        AX_LUA_LIBS([
        ],[AC_MSG_ERROR([Cannot find Lua libs.   Please install lua5.2 liblua5.2-dev])])
    ],[AC_MSG_ERROR([Cannot find Lua includes.  Please install lua5.2 liblua5.2-dev])])
],[AC_MSG_ERROR([Cannot find Lua interpreter.   Please install lua5.2 liblua5.2-dev])])

AX_LIB_SQLITE3(3.0.0)

if test "x\$SQLITE3_VERSION" = "x"; then 
  AC_MSG_ERROR(['libsqlite-dev' version >= 3.0.0 is required.  Please install libsqlite-dev.]) 
fi 

# Check for Geos library
AX_LIB_GEOS(3.0.0)
if test "x\$GEOS_VERSION" = "x"
then
  AC_MSG_ERROR(['geos' version >= 3.0.0 is required.  Please install geos.]);
fi

# spatialite needed for admin info
PKG_CHECK_MODULES([LIBSPATIALITE], [spatialite >= 3.0.0], , AC_MSG_ERROR(['libspatialite-dev' version >= 3.0.0 is required.  Please install libspatialite-dev.]))

# check pkg-config packaged packages.
PKG_CHECK_MODULES([DEPS], [protobuf >= 2.4.0 libprime_server >= 0.3.4 libcurl >= 7.35.0])

# optionally enable coverage information
CHECK_COVERAGE

AC_CONFIG_FILES([Makefile libvalhalla.pc])

# Debian resets this to no, but this break both Spot and the libtool
# test suite itself.  Instead of requiring developer to install a
# non-patched version of Libtool on any Debian they use, we just
# cancel the effect of Debian's patch here.
# see: http://git.lrde.epita.fr/?p=spot.git;a=commitdiff;h=0e74b76521341f670f6b76f8ef24a6dcf6e3813b
link_all_deplibs=yes
link_all_deplibs_CXX=yes

AC_OUTPUT
EOF

popd


local.sh
#!/bin/bash
set -e

#the main place for info on how to do this is here: http://packaging.ubuntu.com/html/index.html
#section 2. launchpad here: http://packaging.ubuntu.com/html/getting-set-up.html
#section 6. packaging here: http://packaging.ubuntu.com/html/packaging-new-software.html

VERSION=$(cat version)
RELEASES=$(cat releases)
declare -A boost=( ["trusty"]="1.54" ["vivid"]="1.55" ["wily"]="1.58" ["xenial"]="1.58" )

#get the pre install hooks
sudo hooks/D10addppa

#get a bunch of stuff we'll need to  make the packages
sudo apt-get install -y dh-make dh-autoreconf bzr-builddeb pbuilder debootstrap devscripts distro-info
#get the stuff we need to build the software
sudo apt-get install -y autoconf automake pkg-config libtool make gcc g++ lcov
sudo apt-get install -y $(grep -F Build-Depends debian/control | sed -e "s/(.*),//g" -e "s/,//g" -e "s/^.*://g" -e "s/BOOST_VERSION/${boost[${DISTRIB_CODENAME}]}/g")

#tell bzr who we are
DEBFULLNAME="valhalla"
DEBEMAIL="valhalla@mapzen.com"
bzr whoami "${DEBFULLNAME} <${DEBEMAIL}>"
source /etc/lsb-release

#versioned package name
PACKAGE="$(if [[ "${1}" == "--versioned-name" ]]; then echo libvalhalla${VERSION}; else echo libvalhalla; fi)"

######################################################
#SEE IF WE CAN BUILD THE PACKAGE FOR OUR LOCAL RELEASE
rm -rf local_build
mkdir local_build
#get pieces of code into the form bzr likes
./prepare.sh ${VERSION} local_build/${PACKAGE}
pushd local_build
tar pczf ${PACKAGE}.tar.gz ${PACKAGE}
rm -rf ${PACKAGE}

#start building the package, choose l(ibrary) for the type
bzr dh-make ${PACKAGE} ${VERSION} ${PACKAGE}.tar.gz << EOF
l

EOF

#bzr will make you a template to fill out but who wants to do that manually?
rm -rf ${PACKAGE}/debian
cp -rp ../debian ${PACKAGE}
if [[ "${1}" == "--versioned-name" ]]; then
	echo -e "libvalhalla${VERSION} (${VERSION}-0ubuntu1~${DISTRIB_CODENAME}1) ${DISTRIB_CODENAME}; urgency=medium\n" > ${PACKAGE}/debian/changelog
	for p in $(grep -F Package ${PACKAGE}/debian/control | sed -e "s/.*: //g"); do
		for ext in .dirs .install; do
			mv ${PACKAGE}/debian/${p}${ext} ${PACKAGE}/debian/$(echo ${p} | sed -e "s/valhalla/valhalla${VERSION}/g" -e "s/valhalla${VERSION}\([0-9]\+\)/valhalla${VERSION}-\1/g")${ext}
		done
	done
	sed -i -e "s/\([b| ]\)valhalla/\1valhalla${VERSION}/g" -e "s/valhalla${VERSION}\([0-9]\+\)/valhalla${VERSION}-\1/g" ${PACKAGE}/debian/control
else
	echo -e "libvalhalla (${VERSION}-0ubuntu1~${DISTRIB_CODENAME}1) ${DISTRIB_CODENAME}; urgency=medium\n" > ${PACKAGE}/debian/changelog
fi
sed -i -e "s/BOOST_VERSION/${boost[${DISTRIB_CODENAME}]}/g" ${PACKAGE}/debian/control
curl https://raw.githubusercontent.com/valhalla/valhalla-docs/master/release-notes.md 2>/dev/null | sed -e "s/^##/*/g" -e "s/^\(.\)/  \1/g" >> ${PACKAGE}/debian/changelog
echo -e "\n -- ${DEBFULLNAME} <${DEBEMAIL}>  $(date -u +"%a, %d %b %Y %T %z")" >> ${PACKAGE}/debian/changelog

#add the stuff to the bzr repository
pushd ${PACKAGE}
bzr add debian
bzr commit -m "Packaging for ${VERSION}-0ubuntu1."

#build the packages
bzr builddeb -- -us -uc -j$(nproc)
popd
popd
######################################################

